### PR TITLE
Use an atomic operation to calculate Issue.Index

### LINKED ISSUE #7887: models/issue.go NewIssue() could fail due to duplicate key
The current code for `models/issue.go` roughly does:

```
func getMaxIndexOfIssue(e Engine, repoID int64) (int64, error) {
    var (
        maxIndex int64
        has      bool
        err      error
    )

    has, err = e.SQL("SELECT COALESCE((SELECT MAX(`index`) FROM issue WHERE repo_id = ?),0)", repoID).Get(&maxIndex)

   [...]

    return maxIndex, nil
}

func newIssue(e *xorm.Session, doer *User, opts NewIssueOptions) (err error) {
	opts.Issue.Title = strings.TrimSpace(opts.Issue.Title)

	maxIndex, err := getMaxIndexOfIssue(e, opts.Issue.RepoID)
	if err != nil {
		return err
	}
	opts.Issue.Index = maxIndex + 1

   [...]

	if _, err = e.Insert(opts.Issue); err != nil {
		return err
	}

   [...]
}
```

It's possible that under heavy loads (e.g. a small server) two sessions may attempt to use the same index number to create issues; one of the two will fail due to a race condition:

1. User **A** submits the form creating an issue for repo abc.
2. User **B** submits the form creating another issue for repo abc.
3. At `issue.go`, in the context of user **A**'s request, `getMaxIndexOfIssue()` returns `99`. It will be used at step *5*.
4. At `issue.go`, in the context of user **B**'s request, `getMaxIndexOfIssue()` also returns `99`. because at the moment that's the latest issue for the repo. It will be used at step *6*.
5. At `issue.go`, in the context of user **A**'s request, `newIssue()` inserts **A**'s issue with `Index: 100`.
6. At `issue.go`, in the context of user **B**'s request, `newIssue()` attempts to insert **B**'s issue with `Index: 100` and fails with `duplicate key error`, because `100` was used at step *5*.

This will of course depend on the database locking model.

I think the best way to handle this is to use a mutex, thus covering most scenarios.

### PR DESCRIPTION
**Pull request by guillep2k**
Please notice that I've referenced the master version of xorm (xorm.io/builder,github.com/go-xorm/xorm) to build this. I'm not sure if I did it correctly.

Closes #7887 

This PR takes advantage of the [new feature](https://github.com/go-xorm/xorm/pull/1401) supported by xorm of building the `INSERT` from a `SELECT`.

This is the new code built by xorm (SQLite):

```
2019/08/22 12:21:16 .../xorm/session_raw.go:178:exec() [I] [SQL] INSERT INTO `issue` (`repo_id`, `poster_id`, `original_author`, `original_author_id`, `name`, `content`, `milestone_id`, `priority`, `is_closed`, `is_pull`, `num_comments`, `ref`, `deadline_unix`, `created_unix`, `updated_unix`, `closed_unix`, `is_locked`, `index`) SELECT ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, coalesce(MAX(`index`),0)+1 FROM `issue` WHERE (repo_id=?) []interface {}{1, 1, "", 0, "This is a pull title", "", 0, 0, false, true, 0, "", 0, 1566487276, 1566487276, 0, false, 1} - took: 1.240578ms
```

Instead of:
```
2019/08/22 16:39:39 .../xorm/session_raw.go:83:queryRows() [I] [SQL] SELECT COALESCE((SELECT MAX(`index`) FROM issue WHERE repo_id = ?),0) []interface {}{10} - took: 67.98µs
2019/08/22 16:39:39 .../xorm/session_raw.go:178:exec() [I] [SQL] INSERT INTO `issue` (`repo_id`,`index`,`poster_id`,`original_author`,`original_author_id`,`name`,`content`,`milestone_id`,`priority`,`is_closed`,`is_pull`,`num_comments`,`ref`,`deadline_unix`,`created_unix`,`updated_unix`,`closed_unix`,`is_locked`) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?) []interface {}{10, 2, 12, "", 0, "create a success pr", "", 0, 0, false, true, 0, "", 0, 1566502779, 1566502779, 0, false} - took: 521.156µs
```

This will prevent the possible race condition mentioned in #7848 by making the index number calculated in the same operation as the insert.

A thread with an alternate solution and its discussion is: #7898,

### CODE REVIEW

**Comment by guillep2k** [comment]
I've just noticed that xorm re-reads the calculated value for me, so calling `getIssueByID()` is unnecessary. I'll update this PR in a minute.

**Comment by guillep2k** [comment]
*Rebased due to changes in `go.mod` at `master`.

**Comment by sapk** [comment]
I haven't tested it but does it select the max index for the specific repo id or across all repo ?

**Review by sapk — APPROVED** [review]
Looking at the generated SQL request from xorm repo, it should be good.

**Comment by guillep2k** [comment]
Something wrong with the issue/comment template? Mmm....

**Comment by sapk** [comment]
On a side note: the concurrency problem exist also for PR since they used GetMaxIndexOfIssue. But first let validate this PR. :smile:

File: models/issue.go [hunk 1]
```diff
@@ -1109,10 +1103,19 @@ func newIssue(e *xorm.Session, doer *User, opts NewIssueOptions) (err error) {
 	}
 
 	// Milestone and assignee validation should happen before insert actual object.
-	if _, err = e.Insert(opts.Issue); err != nil {
+	if _, err := e.SetExpr("`index`", "coalesce(MAX(`index`),0)+1").
```
Comment (lafriks) [inline models/issue.go:1106]: Is this really supported in all db types?

Comment (sapk) [reply models/issue.go:1106]: yes it should

Comment (guillep2k) [reply models/issue.go:1106]: @lafriks Check my comment about DB support here: https://github.com/go-gitea/gitea/pull/7898#issuecomment-523685648

As far as the syntax and the functionality, they are supported. About the atomicity, in theory it's there, but It's very difficult to test. There some WIP about that here #7950 and here #7959.

Comment (sapk) [reply models/issue.go:1106]: To validate there is an integration `TestAPICreateIssue` that validate the creation across all db types.

**Comment by guillep2k** [comment]
> 
> 
> On a side note: the concurrency problem exist also for PR since they used GetMaxIndexOfIssue. But first let validate this PR. 😄

@sapk In fact, that value is being ignored (it always was). I might PR for that later.

**Comment by guillep2k** [comment]
@typeless , @sapk, @lunny 
The **good news** is that I've managed to create a test that can make this fail.
The **bad news** is that I've managed to create a test that can make this fail.

See #7959.

Sadly, for solving our duplicate index problem, the `INSERT ... SELECT` method only works with SQLite. All MySQL, MSSQL and PostgreSQL fail one way or another:

**MSSQL**: `insert successfully but not returned id` (internally it's `Cannot insert duplicate key row in object 'dbo.issue' with unique index 'UQE_issue_repo_index'.`).
**MySQL**: `Deadlock found when trying to get lock; try restarting transaction`.
**MySQL8**: `Deadlock found when trying to get lock; try restarting transaction`.
**PostgreSQL**: `pq: duplicate key value violates unique constraint "UQE_issue_repo_index"`.

With MySQL, a lock is in place but somehow creates a deadlock. With MSSQL and PostgreSQL, if there's a lock, it is not kept from the `SELECT` part into the `INSERT` part.

So, we're back to square one regarding this, and the only viable solution at the moment seems to retry the insert, as @lunny proposed.

**Comment by guillep2k** [comment]
Close due to failing tests on #7959

**Comment by lafriks** [comment]
Did you try SetExpr with full select max?

**Comment by guillep2k** [comment]
> 
> 
> Did you try SetExpr with full select max?

@lafriks Yes, that's #7959. I've branched this to add the test. I didn't want to add the test here for performance reasons (the integration tests would have a lot of time added to them for everybody).

**Comment by lunny** [comment]
I don't that will only supports sqlite3 because I have added tests on https://github.com/go-xorm/xorm/pull/1401 and it will be executed on sqlite/mysql/postgres/mssql both on circleci and drone.
https://cloud.drone.io/go-xorm/xorm/67 and https://circleci.com/gh/go-xorm/xorm/952?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by guillep2k** [comment]
> I don't that will only supports sqlite3 because I have added tests on go-xorm/xorm#1401 and it will be executed on sqlite/mysql/postgres/mssql both on circleci and drone.

@lunny I'm sorry, of course the statement works well in all databases. Your modification generates the right code and it passed all the existing tests on Gitea.

However, it does **not** fix #7887. It does not work if multiple users attempt to insert **at the same time**. There are all kinds of concurrency problems. Please see my test in #7959 for details.

**Pull request closed without merging**